src/index.ts

```diff
@@ -0,0 +1,542 @@
+import axios, { Method } from 'axios'
+import { Agent } from 'https'
+import {
+    Device,
+    DeviceTemperatureOffset,
+    Home,
+    Me,
+    MobileDeviceSettings,
+    Weather,
+} from './types'
+
+const EXPIRATION_WINDOW_IN_SECONDS = 300
+
+const tado_auth_url = 'https://auth.tado.com'
+const tado_url = 'https://my.tado.com'
+const oauth_path = '/oauth/token'
+const tado_config = {
+    client: {
+        id: 'tado-web-app',
+        secret: 'wZaRN7rpjn3FoNyF5IFuxg9uMzYJcvOoQ8QWiIqS3hfk6gLhVlG57j5YNoZL2Rtc',
+    },
+    auth: {
+        tokenHost: tado_auth_url,
+    },
+}
+
+const client = new ResourceOwnerPassword(tado_config)
+
+export class Tado {
+    private _httpsAgent: Agent
+    private _accessToken: AccessToken | null
+    private _username: string
+    private _password: string
+
+    constructor(username: string, password: string) {
+        this._username = username
+        this._password = password
+        this._httpsAgent = new Agent({ keepAlive: true })
+    }
+
+    async _login() {
+        const tokenParams = {
+            username: this._username,
+            password: this._password,
+            scope: 'home.user',
+        }
+
+        try {
+            this._accessToken = await client.getToken(tokenParams)
+        } catch (error) {
+            throw error
+        }
+    }
+
+    async _refreshToken() {
+        if (!this._accessToken) {
+            await this._login()
+        }
+
+        if (!this._accessToken) {
+            throw new Error(`No access token available, even after login in.`)
+        }
+
+        // If the start of the window has passed, refresh the token
+        const shouldRefresh = this._accessToken.expired(
+            EXPIRATION_WINDOW_IN_SECONDS
+        )
+
+        if (shouldRefresh) {
+            try {
+                this._accessToken = await this._accessToken.refresh()
+            } catch (error) {
+                this._accessToken = null
+                await this._login()
+            }
+        }
+    }
+
+    async login(username: string, password: string) {
+        this._username = username
+        this._password = password
+        await this._login()
+    }
+
+    async apiCall<R, T = any>(
+        url: string,
+        method: Method = 'get',
+        data?: T
+    ): Promise<R> {
+        await this._refreshToken()
+
+        let callUrl = tado_url + url
+        if (url.includes('https')) {
+            callUrl = url
+        }
+        const request = {
+            url: callUrl,
+            method: method,
+            data: data,
+            headers: {
+                Authorization:
+                    'Bearer ' + this._accessToken?.token.access_token,
+            },
+            httpsAgent: this._httpsAgent,
+        }
+        if (method !== 'get' && method !== 'GET') {
+            request.data = data
+        }
+        const response = await axios(request)
+
+        return response.data as R
+    }
+
+    getMe(): Promise<Me> {
```

Comment:
You can see I removed the `async` key work to return the promise and avoid aving a nested Promise as the end (since we don't await)